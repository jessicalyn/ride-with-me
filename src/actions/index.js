export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const storeCities = (cities) => ({
  type: 'STORE_CITIES',
  cities
})
