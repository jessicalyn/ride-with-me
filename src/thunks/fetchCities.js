import { hasError, storeCities } from '../actions';
import { fetchOptionsCreator } from '../utils/fetchOptionsCreator';

export const fetchCities = () => {
  return async (dispatch) => {
    const url = `${process.env.REACT_APP_BACKEND_URL}/graphql/`
    try {
      const searchParams = '{ allCities{name} }'
      const options = fetchOptionsCreator('POST', searchParams)
      console.log(options)
      const response = await fetch(url, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      const cities = result.data.allCities
      dispatch(storeCities(cities))
    } catch(error) {
      dispatch(hasError(error.message))
    }
  }
}