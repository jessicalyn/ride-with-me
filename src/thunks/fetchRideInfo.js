import { hasError, storeRides, isLoading } from '../actions';
import { fetchOptionsCreator } from '../utils/fetchOptionsCreator';

export const fetchRideInfo = () => {
  return async (dispatch) => {
    const url = 'https://ride-with-me-backend.herokuapp.com/graphql/?query='
    try {
      dispatch(isLoading(true))
      const searchParams = '{ availableRides{id status availableSeats price description mileage travelTime driver{id uuid firstName}ridepassengerSet{passenger{id firstName}}endCity{id name}startCity{id name}} }'
      const options = fetchOptionsCreator('POST', searchParams)
      const response = await fetch(url, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      const rides = result.data.availableRides
      dispatch(isLoading(false))
      dispatch(storeRides(rides))
    } catch(error) {
      dispatch(isLoading(false))
      dispatch(hasError(error.message))
    }
  }
}