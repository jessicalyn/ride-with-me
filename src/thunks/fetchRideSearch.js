import { hasError, isLoading, storeSearchResults } from '../actions';
import { fetchOptionsCreator } from '../utils/fetchOptionsCreator';

export const fetchRideSearch = (queryVariables) => {
  return async (dispatch) => {
    const url = 'https://ride-with-me-backend.herokuapp.com/graphql/?query='
    try {
      dispatch(isLoading(true))
      const options = fetchOptionsCreator('POST', queryVariables)
      const response = await fetch(url, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      dispatch(isLoading(false))
      dispatch(storeSearchResults(result.data.searchRidesByCities))
    } catch(error) {
      dispatch(isLoading(false))
      dispatch(hasError(error.message))
    }
  }
}