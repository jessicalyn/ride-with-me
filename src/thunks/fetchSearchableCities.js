import { hasError, storeSearchableCities, isLoading } from '../actions';
import { fetchOptionsCreator } from '../utils/fetchOptionsCreator';

export const fetchSearchableCities = () => {
  return async (dispatch) => {
    const url = 'https://ride-with-me-backend.herokuapp.com/graphql/'
    try {
      dispatch(isLoading(true))
      const searchParams = '{ searchableCities{endCities{id name}startCities{id name}} }'
      const options = fetchOptionsCreator('POST', searchParams)
      const response = await fetch(url, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      const cities = Object.values(result.data.searchableCities)
      console.log("cities", cities)
      // const startCities = result.data.searchableCities.startCities
      // const endCities = result.data.searchableCities.endCities
      // console.log(startCities)
      // console.log(endCities)
      dispatch(isLoading(false))
      dispatch(storeSearchableCities(cities))
    } catch(error) {
      dispatch(isLoading(false))
      dispatch(hasError(error.message))
    }
  }
}