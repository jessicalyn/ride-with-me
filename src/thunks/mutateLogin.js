import { hasError, loginUser, isLoading } from '../actions';

export const mutateLogin = (variables) => {
  return async (dispatch) => {
    const url = 'https://ride-with-me-backend.herokuapp.com/graphql/?query=mutation($email:String!,$firstName:String!,$lastName:String!,$imageUrl:String!){loginUser(email:$email,firstName:$firstName,lastName:$lastName,imageUrl:$imageUrl){user{id,firstName,lastName,email,uuid,imageUrl}}}&variables='
    try {
      dispatch(isLoading(true))
      const urlVariables = url + variables
      const options = { method: 'POST' }
      const response = await fetch(urlVariables, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      const user = result.data.loginUser.user
      
      console.log("mutateLogin", user)
      dispatch(isLoading(false))
      dispatch(loginUser(user))
    } catch(error) {
      dispatch(isLoading(false))
      dispatch(hasError(error.message))
    }
  }
}