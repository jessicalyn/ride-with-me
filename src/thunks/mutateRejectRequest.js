import { hasError, isLoading } from '../actions';

export const mutateRejectRequest = (url) => {
  return async (dispatch) => {
    try {
      dispatch(isLoading(true))
      const options = { method: 'POST' }
      const response = await fetch(url, options)
      if (!response.ok) {
        throw Error (response.statusText)
      }
      const result = await response.json()
      console.log("mutateUser", result)
      dispatch(isLoading(false))
      return result
    } catch(error) {
      dispatch(isLoading(false))
      dispatch(hasError(error.message))
    }
  }
}