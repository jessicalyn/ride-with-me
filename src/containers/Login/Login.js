import React, { Component } from 'react';
import { connect } from 'react-redux';
import { GoogleLogin } from 'react-google-login';
import { mutateLogin } from '../../thunks/mutateLogin';
import { Redirect } from 'react-router-dom';
import { loginUser } from '../../actions';

export class Login extends Component {
  constructor() {
    super();
    this.state = {
      firstName: "",
      lastName: "",
      email: "",
      googleId: ""
    }
  }

  googleLogin = (response) => {
    console.log(response)
    const { tokenId, profileObj } = response
    this.setState({ firstName: profileObj.familyName, lastName: profileObj.givenName, email: profileObj.email, googleId: tokenId })
    this.userLogin()
  }
  
  userLogin = () => {
    const { firstName, lastName, email } = this.state
    const variables = `{"email":"${email}","firstName":"${firstName}","lastName":"${lastName}"}`
    this.props.mutateLogin(variables)
  }

  tryAgain = (response) => {
    console.log("DELETE THE FAKE USER")
    const user = { 
      email: "jessicalynhansen@gmail.com", 
      firstName: "Hansen",
      id: "12",
      lastName: "Jessica",
      uuid: "a710eedc-819c-11e9-8f08-5a5d1f8d1912" 
    }
    this.props.loginUser(user)
    console.log("google login didn't work")
  }

  render() {
    return(
      <div className="containers">
        <h2>Please Login to Continue</h2>
        <p>Signin with Google to Create an Account or Login</p>
        <GoogleLogin
          clientId="716984648143-q0uo1263pr96hegug4iqlotqslidqu9p.apps.googleusercontent.com"
          buttonText="Login"
          onSuccess={this.googleLogin}
          onFailure={this.tryAgain}
          cookiePolicy={'single_host_origin'}
        />
        { this.props.user && <Redirect to='/' /> }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  mutateLogin: (variables) => dispatch(mutateLogin(variables)),
  loginUser: (user) => dispatch(loginUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(Login)