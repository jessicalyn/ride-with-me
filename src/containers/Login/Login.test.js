import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import React from 'react';
import { shallow } from 'enzyme';
import { mutateLogin } from '../../thunks/mutateLogin';
import { loginUser } from '../../actions';

jest.mock('../../thunks/mutateLogin')
jest.mock('../../actions')

describe('Login Container', () => {
  describe('Login', () => {
    let wrapper

    const mockState = {
      firstName: "",
      lastName: "",
      email: "",
      googleId: "",
      imageUrl: ""
    }
    const mockUser = {id: 1, name: "Jessica"}
    let mockMutateLogin
    let mockLoginUser

    beforeEach(() => {
      mockMutateLogin = jest.fn()
      mockLoginUser = jest.fn()
      wrapper = shallow(
        <Login 
          user={mockUser}
          mutateLogin={mockMutateLogin}
          loginUser={mockLoginUser}
        />
      )
    })

    it('should match the snapshot', () => {
      expect(wrapper.debug()).toMatchSnapshot()
    })

    it('should have proper default state', () => {
      expect(wrapper.state()).toEqual(mockState)
    })

    it('should set state on googleLogin', () => {
      const mockResponse = {
        tokenId: 234314,
        profileObj: {
          familyName: "Hansen",
          givenName: "Hansen",
          email: "Hansen",
          imageUrl: "www.image.com"
        }
      }
      wrapper.instance().googleLogin(mockResponse)

      expect(wrapper.state()).toEqual({
        lastName: "Hansen",
        firstName: "Hansen",
        email: "Hansen",
        googleId: 234314,
        imageUrl: "www.image.com"
      })
    })

    it('should invoke userLogin when googleLogin is called', () => {
      const mockResponse = {
        tokenId: 234314,
        profileObj: {
          familyName: "Hansen",
          givenName: "Hansen",
          email: "Hansen",
          imageUrl: "www.image.com"
        }
      }
      const spy = jest.spyOn(wrapper.instance(), 'userLogin')

      wrapper.instance().googleLogin(mockResponse)

      expect(spy).toHaveBeenCalled()
    })

    it('should invoke mutateLogin when userLogin is called', () => {
      wrapper.instance().userLogin()

      expect(mockMutateLogin).toHaveBeenCalled()
    })

  })

  describe('mapStateToProps', () => {
    it('should return the expected props to state', () => {

      const mockState = {
        user: {id: 1, name: "Jessica"},
        fakeState: "Not real state to return"
      }
      const expected = {
        user: {id: 1, name: "Jessica"}
      }
  
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should dispatch mutateLogin with variables', () => {
      const mockVariables = {"email":"email","firstName":"firstName","lastName":"lastName","imageUrl":"imageUrl"}
      const mockDispatch = jest.fn()
      const actionToDispatch = mutateLogin(mockVariables)
      const mappedProps = mapDispatchToProps(mockDispatch)
  
      mappedProps.mutateLogin(mockVariables)
  
      expect(mockDispatch).toHaveBeenCalled(actionToDispatch)
    })

    it('should dispatch loginUser with user', () => {
      const mockUser = {id: 1, name: "Jessica"}
      const mockDispatch = jest.fn()
      const actionToDispatch = loginUser(mockUser)
      const mappedProps = mapDispatchToProps(mockDispatch)
      
      mappedProps.loginUser(mockUser)

      expect(mockDispatch).toHaveBeenCalled(actionToDispatch)      
    })
  })
  
})