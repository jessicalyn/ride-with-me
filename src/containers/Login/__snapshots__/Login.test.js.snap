// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Container Login should match the snapshot 1`] = `
"<div className=\\"containers\\">
  <h2>
    Please Login to Continue
  </h2>
  <p>
    Signin with Google to Create an Account or Login
  </p>
  <e clientId=\\"716984648143-q0uo1263pr96hegug4iqlotqslidqu9p.apps.googleusercontent.com\\" buttonText=\\"Login\\" onSuccess={[Function]} onFailure={[Function]} cookiePolicy=\\"single_host_origin\\" type=\\"button\\" tag=\\"button\\" scope=\\"profile email\\" accessType=\\"online\\" prompt=\\"\\" fetchBasicProfile={true} isSignedIn={false} uxMode=\\"popup\\" disabledStyle={{...}} icon={true} theme=\\"light\\" onRequest={[Function: onRequest]} jsSrc=\\"https://apis.google.com/js/api.js\\" />
  <Redirect to=\\"/\\" />
</div>"
`;
